File: app/src/main/java/com/sakuna63/tumbin/application/di/component/ApplicationComponent.java
package com.sakuna63.tumbin.application.di.component;

import com.sakuna63.tumbin.application.di.module.ApiModule;
import com.sakuna63.tumbin.application.di.module.ApplicationModule;
import com.sakuna63.tumbin.application.di.module.AuthenticationModule;
import com.sakuna63.tumbin.application.misc.AccountManager;
import com.sakuna63.tumbin.data.api.TumblrService;
import com.sakuna63.tumbin.application.contract.presenter.login.OauthHelper;

import javax.inject.Singleton;

import dagger.Component;
import io.realm.RealmConfiguration;

@Singleton
@Component(modules = {ApplicationModule.class, AuthenticationModule.class, ApiModule.class})
public interface ApplicationComponent {
    AccountManager accountManger();

    OauthHelper oauthHelper();

    RealmConfiguration realmConfiguration();

    TumblrService tumblrService();
}

File: app/src/main/java/com/sakuna63/tumbin/application/di/component/ApplicationComponent.kt
package com.sakuna63.tumbin.application.di.component

import com.sakuna63.tumbin.application.di.module.ActivityModule
import com.sakuna63.tumbin.application.di.module.ApiModule
import com.sakuna63.tumbin.application.di.module.ApplicationModule
import com.sakuna63.tumbin.application.di.module.LoginPresenterModule
import com.sakuna63.tumbin.application.di.module.PostsPresenterModule
import dagger.Component
import okhttp3.OkHttpClient
import javax.inject.Named
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(ApplicationModule::class, ApiModule::class))
interface ApplicationComponent {
    fun plus(module: ActivityModule): ActivityComponent

    fun plus(module: ActivityModule, loginModule: LoginPresenterModule): LoginComponent

    fun plus(module: ActivityModule, postsModule: PostsPresenterModule): HomeComponent

    @Named(ApiModule.NAME_UNSAFE_CLIENT)
    fun unsafeOkHttpClient(): OkHttpClient
}
